tendermint/src/lite/verifier.rs

```diff
@@ -284,42 +239,100 @@ pub fn verify_header<TS, SH, C, L, S, R>(
 ) -> Result<(), Error>
 where
     TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
     L: TrustThreshold,
-    S: Store<TrustedState = TS>,
     R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
-    C: Commit,
+    S: Store<TrustedState = TS>,
+{
+    // Fetch the latest state and ensure it hasn't expired.
+    // Note we only check for expiry once in this
+    // verify_and_update_bisection function since we assume the
+    // time is passed in and we don't access a clock internall.
+    // Thus the trust_period must be long enough to incorporate the
+    // expected time to complete this function.
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
+
+    // TODO: consider fetching the header we're trying to sync to and
+    // checking that it's time is less then `now + X` for some small X.
+    // If not, it means that either our local clock is really slow
+    // or the blockchains BFT time is really wrong.
+    // In either case, we should probably raise an error.
+    // Note this would be stronger than checking that the untrusted
+    // header is within the trusting period, as it could still diverge
+    // significantly from `now`.
+
+    // inner recursive function which assumes
+    // trusting_period check is already done.
+    _verify_and_update_bisection(untrusted_height, trust_threshold, req, store)
+}
+
+// inner recursive function for verify_and_update_bisection.
+// see that function's docs.
+fn _verify_and_update_bisection<TS, SH, C, L, R, S>(
+    untrusted_height: Height,
+    trust_threshold: &L,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
     SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
+    S: Store<TrustedState = TS>,
 {
-    // Check if we already trusted a header at the given height and it didn't expire:
-    if let Ok(ts2) = store.get(height) {
-        is_within_trust_period(ts2.last_header().header(), trusting_period, now)?
-    }
-
-    // We haven't trusted a header at given height yet. Request it:
-    let sh2 = req.signed_header(height)?;
-    let sh2_vals = req.validator_set(height)?;
-    verify(&sh2, &sh2_vals)?;
-    is_within_trust_period(sh2.header(), trusting_period, now)?;
-
-    // Get the highest trusted header with height lower than sh2's.
-    let sh1_trusted = store.get_smaller_or_equal(height)?;
-    can_trust_bisection(
-        &sh1_trusted,
-        &sh2,
+    // this get is redundant the first time.
+    // TODO: possibly refactor so this func takes and returns
+    // trusted_state.
+    let trusted_state = store.get(Height::from(0))?;
```

Comment:
Yeh, this is a bit inefficient as we keep getting the trusted state from the store for all pivots that we just added in L#311. It would be maybe better if the trusted state is passed in as parameter. And easier to read.

The other question is about store implementation. I expect there is some background routine that runs periodically and purges expired headers from the store. This would interfere with bisection runs and needs to be analyzed as getting a trusted header from the store might fail.

Reply:
Yes I thought it would be easiest to read and write first like this. And also figured stores would have caching so it wouldn't be inefficient. But see discussion with xla about not passing in store into these functions https://github.com/interchainio/tendermint-rs/issues/110#issuecomment-569973976

Been thinking about a Client state machine and functions that return TrustedState so that it's the users responsibility to make the store updates but they can't do it wrong.

Pruning should be guaranteed to not interfere - ie. it can only apply outside the trusting period.